Experiment flag for using amp-base-carousel from amp-lightbox-gallery.

sparhami opened the pull request:
There is still more work to do before actually getting this out, but setting up the groundwork.

- Implement methods needed for next/prev on `amp-base-carousel`.
- Change carousel creation / reuse tags and selectors to use
`amp-base-carousel` for the experiment.
- Fix a race between `goToSlide` and `updateUi` which would result in the call
  to `goToSlide` being ignored.

For #21568

cathyxz commented on extensions/amp-base-carousel/0.1/carousel.js:
```diff
@@ -229,6 +227,15 @@ export class Carousel {
 
+    /**
+     * Keeps track of an index that was requested to be scrolled to
+     * prograatically. This is used to make sure that the carousel ends on the
```
Typo: `programmatically`.

cathyxz commented on extensions/amp-base-carousel/0.1/carousel.js:
```diff
@@ -834,11 +844,22 @@ export class Carousel {
+    // If we are still on the same slide, nothing needs to move. If there is
+    // a programmatic scroll to an index, we may need to move things
+    // accordingly.
+    if (this.restingIndex_ == this.currentIndex_ &&
```
Nit: without reading the initialization comments for `current` and `resting` index, it's not immediately clear to me what the difference between these two are.

sparhami replied:
Reworded this a bit, hopefully more clear now,

sparhami commented:
> We chatted offline about this: there could potentially be a race condition where if goToSlide is called in the middle of the user interacting with the carousel, it would directly jump to the requested slide and result in a jarring UX. We should either fix or file / note this behavior.

Added logic to `goToSlide` to bail out if the user is scrolling. Also added a small fix to allow calling advance multiple times in a row before the current index has updated.

cathyxz commented on extensions/amp-base-carousel/0.1/carousel.js:
```diff
@@ -838,6 +869,8 @@ export class Carousel {
    * @private
    */
   resetScrollReferencePoint_(force = false) {
+    this.scrolling_ = false;
```
Should this be after making sure that the user is not in the middle of a drag?

sparhami replied:
I considered it, but we already have something to check for touching. So if they stopped scrolling (no scroll event recently), but they are still touching the screen, I think it would be cleaner to say that there is no scrolling, and interested code can check for touching as well. If they are dragging, we should get scroll events, so that would still count as scrolling.
